### ISSUE TITLE
Counter-intuitive (lack of) side effects from repeated `Require`

### ISSUE THREAD

**Issue by Janno** [coq/coq#19294]
### Description of the problem

In the final file of [bug.zip](https://github.com/user-attachments/files/16032291/bug.zip) , `client_of_client.v` ends up with an obligation tactic value of `fail` that is inherited from `set.v`. This is counter-intuitive because `client_of_client.v` transitively `Require`s `reset.v` through `careful_client.v` as its last `Require`. This should reset the obligation tactic to `idtac`. `Require bug.reset` is also the last thing that happens in `careful_client.v` and `careful_client.v` itself observes the correct value (`idtac`) for the obligation tactic.

I guess that this happens because `reset.v` has already been transitively `Require`d by `client_of_client.v` through `Require bug.requires_reset` before `Require bug.careful_client.v`. I still consider this a bug in terms of my own intuition as well as the impossibly strict discipline this behavior imposes on larger developments. (It took several seasoned coq users a considerable amount of time to track down where our obligation tactic values come from and why they were not reset correctly.)

### Version of Coq where this bug occurs

8.19

**Comment by SkySkimmer** [comment]
That's ndeed how it works. Trying to do otherwise would probably be super inefficient.

>the impossibly strict discipline this behavior imposes on larger developments

This is why require time effects are bad and import time should be used instead.

**Comment by SkySkimmer** [comment]
(you didn't include `requires_reset.v` in the zip, I'm assuming it just does `Require reset`)

**Comment by Janno** [comment]
> (you didn't include requires_reset.v in the zip, I'm assuming it just does Require reset)

That's indeed what it does. I fixed the archive.

> This is why require time effects are bad and import time should be used instead.

What is the policy for `#[global]` effects in Coq itself? Should we be opening bugs for them?

**Comment by SkySkimmer** [comment]
You can open issues about anything you consider a problem.
If the only thing the issue says is "foo uses global which is disrecommended" it's going to be pretty low priority and developers probably won't touch it to avoid having to deal with backward compat issues.
If you the global gives you problems in practice it's more likely to be treated.

**Comment by SkySkimmer** [comment]
For instance if set.v in your original problem is Program/Tactics.v from stdlib we can look into removing the global setting there.
It's probably possible to do a deprecation phase with similar methods as the implicit auto with * in intuition.

**Comment by ejgallego** [comment]
I agree with the view that this is annoying.

Maybe we could have a warning for these kind of problems. I'd suggest:
- taint libraries that perform global effects 
- do a warning if such libraries are used in a possibly non-deterministic configuration (for example imported twice)

What do you folks think?

**Comment by Janno** [comment]
I think that would be an amazing improvement to usability!

**Comment by ejgallego** [comment]
@SkySkimmer what do you think, would it be ok to reopen this bug as a feature request?

**Comment by SkySkimmer** [comment]
Needs to be more specified before we can try to implement something imo but you can open if you want

**Comment by gmalecha** [comment]
I think it could be worth the experiment of simply trying to remove the global side effects to see what impact it has. Several years ago, the number of users of `program` was quite small compared to the full ecosystem, so it *might* be possible to actually remove the effects at this point.

That doesn't solve the broader problem, which I agree with @SkySkimmer is a bit under-specified. For example, if you have a `#[global]` typeclass instance, should it be flagged as one of these effects? Especially `#[global] Hint Extern` since these are actually not removable.

**Comment by Janno** [comment]
My take is that the warning would be most useful (and most actionable) if it tracks *conflicting* global effects. For hints that means `#[global] Hint Resolve` versus `#[global] Remove Hints`, for example. Not `#[global] Hint Extern` since those can never be undone and thus cannot lead to a conflict.

To make this work, I would equip every library with a graph of all the superglobal effects it (immediately) required and where they come from. I think the graph doesn't need to be transitive in `Require`; the edges would just signify that there is an ordering constraint between required modules in the current file. In the example in this issue that would mean `careful_client.v` would record superglobal effects on `Obligation Tactic` from `set.v` and also from `reset.v` and there is an edge from the effect of `set.v` to the effect of `reset.v` to indicate that `careful_client.v` required these in exactly that order.

`client_of_client.v` also builds up a graph like that when performing the two `Require` statements. First, it records the effect on `Obligation Tactic` from `careful_client.v` and it pulls in that file's graph. It then records the effect on `Obligation Tactic` from `reset.v` which adds an edge in the graph from the effect of `careful_client.v` to the effect of `reset.v`. We can now notice a cycle by looking at the edges that are part of the effect coming from `careful_client.v` and looking for `reset.v`.

In the most general case, where `careful_client.v` transitively requires the `reset.v` effect and `client_of_client.v` requires both of these files only transitively, we would have to dig deeper into the graphs of transitively required files. This could obviously be expensive but I am not sure of the complexity. I suspect it would help a lot to tag the recorded effects with a boolean that indicates if there was any ordering constraint so far on that particular effect. Maybe there are few conflicting effects overall.
